process_report/processors/coldfront_fetch_processor.py

```diff
@@ -0,0 +1,104 @@
+import os
+import functools
+import logging
+from dataclasses import dataclass
+
+import requests
+
+from process_report.invoices import invoice
+from process_report.processors import processor
+
+logger = logging.getLogger(__name__)
+logging.basicConfig(level=logging.INFO)
+
+
+CF_ATTR_ALLOCATED_PROJECT_NAME = "Allocated Project Name"
+
+
+@dataclass
+class ColdfrontFetchProcessor(processor.Processor):
+    nonbillable_projects: list[str]
+
+    @functools.cached_property
+    def coldfront_client(self):
+        keycloak_url = os.environ.get("KEYCLOAK_URL", "https://keycloak.mss.mghpcc.org")
+
+        # Authenticate with Keycloak
+        token_url = f"{keycloak_url}/auth/realms/mss/protocol/openid-connect/token"
+        try:
+            r = requests.post(
+                token_url,
+                data={"grant_type": "client_credentials"},
+                auth=requests.auth.HTTPBasicAuth(
+                    os.environ["KEYCLOAK_CLIENT_ID"],
+                    os.environ["KEYCLOAK_CLIENT_SECRET"],
+                ),
+            )
+        except KeyError:
+            logger.error(
+                "Keycloak authentication failed. Please set the environment variables KEYCLOAK_CLIENT_ID and KEYCLOAK_CLIENT_ID"
+            )
+
+        client_token = r.json()["access_token"]
+
+        session = requests.session()
+        headers = {
+            "Authorization": f"Bearer {client_token}",
+            "Content-Type": "application/json",
+        }
+        session.headers.update(headers)
+        return session
+
+    def _get_project_list(self):
+        return self.data[invoice.PROJECT_FIELD].unique()
+
+    def _fetch_coldfront_allocation_api(self, allocation_list):
+        coldfront_api_url = os.environ.get(
+            "COLDFRONT_URL", "https://coldfront.mss.mghpcc.org/api/allocations"
+        )
+        api_query_str = "&".join(
+            [
+                f"attr_{CF_ATTR_ALLOCATED_PROJECT_NAME}={project}"
+                for project in allocation_list
+            ]
+        )
+        r = self.coldfront_client.get(f"{coldfront_api_url}?{api_query_str}")
```

Comment:
so it looks like instead of asking for all allocations, we are going to pass in a giant list of all the allocations in our report. Is this really that much more scalable?

Reply:
@naved001 That's a valid concern. I believe it is pretty scalable up to a certain point.

Each allocation will require 28 characters for `attr_Allocated Project Name=` + 64 characters for max allocation name length <= 100 characters total per allocation query.

Assuming 100 allocations active * 100 per allocation query = 10 000.

At 1byte per KB that's 10kb, which is already much less in size than the complete list of all allocations.

My only concern would be the whether nginx ingress controller is configured or supports that size of URIs out of the box.

@QuanMPhm can you do some edge testing in your openshift cluster with some really high number of allocations? say 100 and then 1000?

Reply:
@knikolla I wrote a script that queried a coldfront deployment running in my local Openshift cluster. When quyerying 100 allocations (i.e `/allocations?allo=1&allo=2&...`), the query took 0.3s. At 1000 allocations, the query took 2.3s. Any more allocations beyond that, and we hit [Django's default query field limit](https://docs.djangoproject.com/en/5.1/ref/settings/#data-upload-max-number-fields).